src/base/bittorrent/portforwarderimpl.h

```diff
@@ -50,14 +50,16 @@ class PortForwarderImpl final : public Net::PortForwarder
     bool isEnabled() const override;
     void setEnabled(bool enabled) override;
 
-    void addPort(quint16 port) override;
-    void deletePort(quint16 port) override;
+    void replacePorts(const QString &profile, QSet<quint16> ports) override;
```

Comment:
Why not just replace "add port" behavior with "set port" one that would combine add/replace logic?
IMO, your suggestion is (slightly) confusing and isn't reliable/consistently implemented (I mostly refer to "profile").

Reply:
The main advantage of profile based is that it doesn't need to remember last used port. IMO it is cumbersome (and error prone) for the caller code have a last used port variable and maintain it.

>IMO, your suggestion is (slightly) confusing and isn't reliable/consistently implemented (I mostly refer to "profile").

I don't get what you mean by "reliable/consistently".
From port forwarding viewpoint, it is natural to see a profile concept. Later I'll make use of it for embedded tracker which is another profile/service.
Having low-level operation to open, close ports is ok but it is unwieldy as mentioned above.

Reply:
> The main advantage of profile based is that it doesn't need to remember last used port

The following too:
>replace current "add port" behavior with the one that just does nothing in case when the given port is already mapped

Reply:
> I don't get what you mean by "reliable/consistently".

I told that it "isn't reliable/consistently". E.g. how it supposed to behave in the following case?
```c++
forwarder->replacePorts("foo", {1234});
forwarder->replacePorts("bar", {1234});
```

Reply:
> > The main advantage of profile based is that it doesn't need to remember last used port
> 
> The following too:
> 
> > replace current "add port" behavior with the one that just does nothing in case when the given port is already mapped

No. Your suggestion still need to decide whether the old port needs to be closed (if new port is different) or left as is (if port hasn't changed).

>I told that it "isn't reliable/consistently". E.g. how it supposed to behave in the following case?

The second port forwarding should fail however I don't think this is worth considering because the port is forwarded successfully via the first call. The outcome is the same with or without this PR.
Also note that this situation is actually user's fault, choosing the same port for different services, nothing good will happen.

>replacePorts sounds low-level too. I would prefer setPorts.

OK, will rename.

Reply:
> So why not prevent user from this issue by preventing the same ports from being contained by different profiles?

Although it turns out that assigning the same port to different services has the same negative effect in other places, so it will cause problems anyway, even if port forwarding is disabled.

Reply:
> > replacePorts sounds low-level too. I would prefer setPorts.
> 
> OK, will rename.

Done.

Reply:
Added logic to handle this situation:
```c++
replacePorts("foo", {1234});
replacePorts("bar", {1234});
removePorts("bar");  // now has no effect on port `1234`
removePorts("foo");  // has effect
```

Reply:
> Added logic to handle this situation:

Hmm... a good solution. So if the user sets two matching ports and then fixes it, forwarding should remain correct.

Reply:
> Added logic to handle this situation:
> 
> ```c++
> replacePorts("foo", {1234});
> replacePorts("bar", {1234});
> removePorts("bar");  // now has no effect on port `1234`
> removePorts("foo");  // has effect
> ```

Unfortunately I cannot locate it in the code. Instead it just silently rejects duplicated port:
https://github.com/qbittorrent/qBittorrent/blob/6674c17e36bcc5a4e2e352323cfd9d3a5654d9c6/src/base/bittorrent/portforwarderimpl.cpp#L89-L90

Reply:
> Added logic to handle this situation:
> 
> ```c++
> replacePorts("foo", {1234});
> replacePorts("bar", {1234});
> removePorts("bar");  // now has no effect on port `1234`
> removePorts("foo");  // has effect
> ```

The following should be also handled in the same way:
```c++
setPorts("foo", {1234});
setPorts("bar", {1234});
removePorts("foo");  // has no effect on port mapping of `1234`, just delete it from "foo" profile
removePorts("bar");  // delete `1234` from "bar" profile and from mapping
```

Reply:
>So it does:

That is expected IMO since port is already taken by `foo`. `bar` cannot own it.

>The following should be also handled in the same way:

That would be more complicated and unnecessary since the port is used/owned by `foo` not `bar`. If `foo` removes the port it should have the effect.

Reply:
> That is expected IMO since port `1234` is already taken by `foo`. `bar` cannot own it.

But this is not reflected in the following place in any way:
```c++
setPorts("bar", {1234});
```

Reply:
The problem is that the following sequence of actions will lead to undefined/incorrect behavior, although I would expect that this should not be the case:
1. Set port of service1 to 1234 and enable forwarding for it;
2. (Mistakenly) set port of service2 to the same as above value (i.e. 1234)  and enable forwarding for it;
3. After some time realize that a mistake has been made and change the port of service1 to different one (e.g. 8888).

Reply:
> > The following should be also handled in the same way:
> 
> That would be more complicated

1. Delete corresponding entry from given profile
2. Check all the profiles for corresponding entries. If there is none, remove port forwarding

Reply:
> The problem is that the following sequence of user actions will lead to undefined/incorrect behavior, although I would expect that this should not be the case:
> 
>     1. Set port of service1 to 1234 and enable forwarding for it;
> 
>     2. (Mistakenly) set port of service2 to the same as above value (i.e. 1234)  and enable forwarding for it;
> 
>     3. After some time realize that a mistake has been made and change the port of service1 to different one (e.g. 8888).

Seems you had a wrong idea of how listening on port works
At step 2 `service2` is trying to listen on the same port `1234` but it will fail: OS will tell you that `1234` is already taken and therefore `service2` is in failed state.
At step 3, `service1` is ok but `service2` is still in failed state due to the previous error until you manually try re-listening on to it. So there is no point to keep port `1234` forwarded when `service1` changed port.
https://stackoverflow.com/a/1694148

Reply:
> At step 3, `service1` is ok but `service2` is still in failed state due to the previous error until you manually try re-listening on to it.

But if this (re-listening) is still done in some way, then the problem with forwarder will remain.

Reply:
>But if this (re-listening) is still done in some way, then the problem with forwarder will remain.

Just like re-listening on port, it must be properly done by invoking port forwarder to open the ports again (which is what current code in webui.cpp is doing).

Reply:
I wonder how your example firewall would handle a case with a duplicate port?

Reply:
> it must be properly done by invoking port forwarder to open the ports again

But this is completely not obvious based only on port forwarder interface. Without knowing the details of the implementation, it may seem obvious that once a certain port has been successfully set for forwarding (namely, this follows from the fact that no error status was returned), then it is still forwarded regardless of errors in other places.

Reply:
I agree that in our very specific case we could simplify it, because it is intended only for internal use and we are confident of how it is used. But then it could be simplified even more, but it's you trying to add some genericness there instead.

Reply:
>But this is completely not obvious based only on port forwarder interface.

OK, I can modify `setPorts` to return a `bool` to indicate whether the operation is successful. If any port was already forwarded by other services or any other errors occurs, it will return `false`.

Reply:
> > But this is completely not obvious based only on port forwarder interface.
> 
> OK, I can modify `setPorts` to return a `bool` to indicate whether the operation is successful. If any port was already forwarded by other services or any other errors occurs, it will return `false`.

It would still look more natural to me if port forwarder allowed the same ports in different profiles, and kept the port forwarded as long as at least one profile referenced it. After all, a duplicate port in this case does not mean "twice forwarded" at all, but only that several "clients" are interested in having it forwarded.
But as I have already mentioned, I no longer insist on anything in this PR.

Reply:
>It would still look more natural to me if port forwarder allowed the same ports in different profiles, and kept the port forwarded as long as at least one profile referenced it. After all, a duplicate port in this case does not mean "twice forwarded" at all, but only that several "clients" are interested in having it forwarded.

OK now I understand your viewpoint. However this "same port" scenario at hand is a fatal error and it is not very clear (to me) that your suggested behavior would be desired. Lets keep the current and if issues arise in the future, then revise it.

Reply:
>OK now I understand your viewpoint. 

👍

> However this "same port" scenario at hand is a fatal error

If you really understand my viewpoint that you understand that it isn't any kind of error for port forwarder itself (if it consider duplicated ports as I described above). It just forwards some ports that someone is interested in forwarding. Thus, it really behaves universally, because it is not based on the belief that profiles are uniquely mapped to network services, and that a certain port, being set simultaneously in several profiles, necessarily means some kind of third-party error.


> Lets keep the current and if issues arise in the future, then revise it.

Ok.